openedx_proversity_reports/reports/time_spent_report.py

```diff
@@ -0,0 +1,239 @@
+"""
+Time spent reports.
+"""
+import logging
+
+from apiclient.discovery import build
+from completion.models import BlockCompletion
+from django.conf import settings
+from django.contrib.auth.models import User
+from google.oauth2 import service_account
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import CourseKey
+
+from openedx_proversity_reports.edxapp_wrapper.get_course_blocks import get_course_blocks
+from openedx_proversity_reports.edxapp_wrapper.get_modulestore import get_modulestore
+
+
+ANALYTICS_API_SCOPES = ['https://www.googleapis.com/auth/analytics.readonly']
+COURSEWARE_URL_PREFIX = '/courseware'
+logger = logging.getLogger(__name__)
+
+
+def get_time_spent_report_data(course_list):
+    """
+    Generates a dict with information about course strcuture and Google Analytics data.
+
+    Args:
+        course_list: A list of valid course ids.
+    Returns:
+        Course structure and Google Anlytics data or an empty dict.
+        {
+            <course_id>: {
+                course_structure: Course structure by subsection.
+                analytics_data: Google Analytics data per course.
+            }
+        }
+    """
+    course_structure = {}
+
+    for course_id in course_list:
+        try:
+            course_key = CourseKey.from_string(course_id)
+        except InvalidKeyError:
+            continue
+
+        analytics_data = fetch_data_from_analytics(course_id)
+
+        # If not GA data, just continue and don't fetch users or course data.
+        if not analytics_data:
+            course_structure[course_id] = {}
+            continue
+
+        # Getting just the first enrolled student in the course, except staff users.
+        enrolled_student = User.objects.filter(
+            courseenrollment__course_id=course_key,
+            courseenrollment__is_active=1,
+            courseaccessrole__id=None,
+            is_staff=0,
+        ).first()
+
+        if not enrolled_student:
+            continue
+
+        usage_key = get_modulestore().make_course_usage_key(course_key)
+        blocks = get_course_blocks(enrolled_student, usage_key)
+        course_block_data = []
+
+        course_blocks = blocks.topological_traversal()
+        course_blocks = list(course_blocks)
```

Comment:
you can merge this line with the previous line

Reply:
Addressed.